libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 		requireLValue(_operation.subExpression(), false);
 	else
 		_operation.subExpression().accept(*this);
-	Type const* subExprType = type(_operation.subExpression());
-	TypeResult result = type(_operation.subExpression())->unaryOperatorResult(op);
-	if (!result)
+	Type const* operandType = type(_operation.subExpression());
+
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = operandType->unaryOperatorResult(op);
+	set<FunctionDefinition const*, ASTNode::CompareByID> matchingDefinitions = operandType->operatorDefinitions(
+		op,
+		*currentDefinitionScope(),
+		true // _unary
+	);
+
+	// Operator can't be both user-defined and built-in at the same time.
+	solAssert(!builtinResult || matchingDefinitions.empty());
+
+	// By default use the type we'd expect from correct code. This way we can continue analysis
+	// of other expressions in a sensible way in case of a non-fatal error.
+	Type const* resultType = operandType;
+
+	FunctionDefinition const* operatorDefinition = nullptr;
+	if (builtinResult)
+		resultType = builtinResult;
+	else if (!matchingDefinitions.empty())
+	{
+		if (matchingDefinitions.size() >= 2)
+		{
+			SecondarySourceLocation secondaryLocation;
+			for (FunctionDefinition const* definition: matchingDefinitions)
+				secondaryLocation.append("Candidate definition:", definition->location());
+
+			m_errorReporter.typeError(
```

Comment:
Maybe but they we'd have to deal with the using statement not being correct. At this point we can assume that they passed all the checks.

Or... wait, is it possible that we'll visit an operation before all the `using` directives are parsed?

Reply:
But anyway, I don't think it's possible at `using for` level because there may be multiple directives defining operators for the same type. You have to process all of them to know if the operator definition is unique or not.

Reply:
Yeah, but we can do that in principle - but yeah, anyways, we won't do it now no matter what.

Reply:
Actually, ok, it is possible. I was assuming that `TypeChecker` sets some annotations on `UsingForDirective` but apparently not. It's probably all set by earlier analysis stages so I guess I could detect this while analyzing the directive.

Reply:
Yeah - but I think we have a lot of analog cases and most of them only throw errors on use - so maybe we just fix that in bulk some time. It's technically breaking for the cases in which we already have this kind of thing.
So yeah, let's keep it as is here for now.

Reply:
Is this actually breaking though? As long as good code keeps compiling and bad code keeps having errors, I think this is something we could change.

If you want, I can create an issue to be done after we merge this PR. It could be a good task for a contributor.

While at it, we could also store the found function in the AST. It's not very efficient that we rescan all `using for` directives we encounter an operator.

Reply:
I don't think it's a task for a contributor :-). Checking for all other situations that may be analog alone is hard. But yeah, an issue would be good.
And well, it's technically breaking. There is good code that compiles and then doesn't after the change. The fact that the good code is only good as long as it has redundant statements doesn't change that really :-). But it's a technicality, so we could get away with it anyways.

Reply:
Actually, looks like it should not be a problem outside of operators. I.e. attached functions are resolved immediately in `using for` and ambiguous definitions are not accepted.

For operators it's different because you can't tell that just looking at a single `using for`. You have multiple functions defining the same operator, being added by separate directives. I guess that's why the check was not originally in `using for`. But yeah, it's doable so I'll change that. I was going to file an issue for it but since it's only for operators and we postponed the release, I can do it properly already in this PR.